text/2020-12-25-selectors.md

```diff
@@ -0,0 +1,242 @@
+    - [Chooser](#chooser)
+  - [Drawbacks](#drawbacks)
+  - [Alternatives](#alternatives)
+  - [Unresolved questions](#unresolved-questions)
+
+<!-- /TOC -->
+
+## Summary
+
+Here is an RFC about more efficient and flexible selectors.
+
+## Motivation
+
+Here are already several issues about the selectors:
+
+- [#886](https://github.com/chaos-mesh/chaos-mesh/issues/886)
+- [#1102](https://github.com/chaos-mesh/chaos-mesh/issues/1102)
+- [#1222](https://github.com/chaos-mesh/chaos-mesh/issues/1222)
+- [#1227](https://github.com/chaos-mesh/chaos-mesh/issues/1227)
+- [#1266](https://github.com/chaos-mesh/chaos-mesh/issues/1266)
+- [#1322](https://github.com/chaos-mesh/chaos-mesh/issues/1322)
+
+In conclusion, issues acquire these features:
+
+- More way to select, like set-based selectors.
+- Select not only Pod resource, but also Service or any other resources.
+- ~~Implement "watching" feature on selectors.~~
+
+> After reconsidering the `watching`, I think it's heavy enough to start a new
+> RFC.
+
+Current Selector implementation is just several methods. It's quite simple but
+not easy to extend.
+
+Before entering detailed design, let's take a look at the current implementation
+of Selector:
+
+Here is an interface called `SelectSpec`, which contains three methods:
+
+```go
+type SelectSpec interface {
+    GetSelector() v1alpha1.SelectorSpec
+    GetMode() v1alpha1.PodMode
+    GetValue() string
+}
+```
+
+All Chaos CRD resource implements this interface; it means all chaos needs the
+selector.
+
+And there are several methods that often used:
+
+```go
+func SelectAndFilterPods(ctx context.Context, c client.Client, r client.Reader, spec SelectSpec, clusterScoped bool, targetNamespace string, allowedNamespaces, ignoredNamespaces string) ([]v1.Pod, error)
+func SelectPods(ctx context.Context, c client.Client, r client.Reader, selector v1alpha1.SelectorSpec, clusterScoped bool, targetNamespace string, allowedNamespaces, ignoredNamespaces string) ([]v1.Pod, error)
+func CheckPodMeetSelector(pod v1.Pod, selector v1alpha1.SelectorSpec) (bool, error)
+```
+
+Notice that "Mode/Value" also affect the final result of select, depends on
+"Mode", it will apply "OnePod", "AllPod", "FixedPod" etc.
+
+On bushiness logic, there are two-part of `SelectSpec`: "SelectorSpec" and
+"Mode/Value". Any resources are filtered/selected by "SelectorSpec" then be
+selected again by "Mode/Value", then we get the final target of chaos injecting.
+
+I think we should split these two-part into different interfaces.
+
+There are several usage for the selector:
+
+- Select certain pods with pod-name.
+- List pods with Kubernetes native label/field selector from target
+  namespace(s), then filter with namespace, annotations, and pod-phase.
+- Check a pod(resource) is matches the selector.
+
+## Detailed design
+
+### SelectSpec
+
+The struct of CRD will not change.
+
+It will add another method for `SelectSpec`:
+
+```go
+type SelectSpec interface {
```

Comment:
Should I implement this interface for custom resources? How do we fetch the custom resource in k8s api? (I'm not much familiar with selector)

Reply:
> Should I implement this interface for custom resources?

You should not implement this interface for custom resources which aims to mirrors pods.
Only Chaos CRD (like NetworkChaos, PodChaos) should implement this interface. And chaos-mesh has already done this by code generation. 

> How do we fetch the custom resource in k8s API?

Here should be an implementation for `Selector`, called `KubernetesSlector`, for example:

```go
struct KubernetesSelector {
    kubeCient client.Client
    labelSelector metav1.LabelSelector
    // other option that could apply on kubernetes native kubeclient API
}

func (it *KubernetesSelector) Select() ([]interface{}, error) {
    listOptions := client.ListOptions{}
    // fill up listOptions
    listOptions.LabelSelector = it.labelSelector
    // ... 
    var result := yourgvk.YourCRDList{}
    err := it.kubeClient.List(ctx, &result, listOptions)
    // ...
    return result
}
```

Reply:
I know... In this case, chaos-mesh can only list local objects like "mirrors resources". So I need to dependent on chaos-mesh api and create resources mirroring my custom resources.

Reply:
No, you do not need Chaos Mesh API when you build your CRD.

I think your mirror resource is extremely like Pod, and obviously, Pod does not depend on Chaos Mesh.

Reply:
But its type should be known by chaos-mesh, shouldn't it? For example, I create a CRD `SomeResource` in package `someone`, then need we write following code in chaos-mesh:

```go
struct SomeResourceSelector {
    kubeCient client.Client
    labelSelector metav1.LabelSelector
    // other option that could apply on kubernetes native kubeclient API
}

func (it *SomeResourceSelector ) Select() ([]interface{}, error) {
    listOptions := client.ListOptions{}
    // fill up listOptions
    listOptions.LabelSelector = it.labelSelector
    // ... 
    var result := someone.SomeResourceList{}
    err := it.kubeClient.List(ctx, &result, listOptions)
    // ...
    return result
}
```

Reply:
> But its type should be known by chaos-mesh, shouldn't it?

If you want these codes should execute in the same binary of Chaos Mesh, then YES.

Reply:
So, in this selector disgin, my understanding is chaos-mesh can provide a new api like `MirrorResource` following.

```go
type MirrorResourceSpec struct {
	// system which provides this mirror
	Provider strring

	// unique id, used as containerID in chaos daemon
	ID string
	
	ChaosDaemonAddress string

	// other nesessary spec
}

type MirrorResourceStatus struct {
	ExecStress struct {
		Instance string
  		StartTime int64;
	}

	// other states ...
}
```

Then other systems can create `MirrorResource` and apply chaos for their resources.

Reply:
No, considering the complex, chaos-mesh will not provide any unnecessary API.

I think it's better to build another binary for handling customize resources as a plugin.🤔 And chaos-mesh will call this plugin with HTTP API. I think I should introduce this(how to make a plugin) in this RFC.

Reply:
Well, looking forward your introduction.